src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +23,451 @@ namespace System
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static Span<T> TrimStart<T>(this Span<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            (int start, _) = Clamp(span, trimElement, true, false);
+            return span.Slice(start);
+        }
+
+        /// <summary>
+        /// Removes all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static Span<T> TrimEnd<T>(this Span<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            (_, int length) = Clamp(span, trimElement, false, true);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Removes all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> Trim<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            (int start, int length) = Clamp(span, trimElement, true, true);
+            return span.Slice(start, length);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimStart<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            (int start, _) = Clamp(span, trimElement, true, false);
+            return span.Slice(start);
+        }
+
+        /// <summary>
+        /// Removes all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimEnd<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            (_, int length) = Clamp(span, trimElement, false, true);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Delimits all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        /// <param name="trimStart">If set to True, trims the start.</param>
+        /// <param name="trimEnd">If set to True, trims the end.</param>
+        private static (int start, int length) Clamp<T>(in ReadOnlySpan<T> span, in T trimElement, bool trimStart, bool trimEnd)
```

Comment:
Will `in` on the ROS block enregistration; or is that only `ref`?

Reply:
I looked at the disassembly for both styles and it looks exactly the same:
```
            int start = ClampViaInStart(in span, trimElement); // Internal call stack uses "in"
00007FFF013A1824  lea         rdx,[rbp+90h]  
00007FFF013A182B  mov         rcx,qword ptr [rbp+88h]  
00007FFF013A1832  call        00007FFF013A11C8  
00007FFF013A1837  mov         dword ptr [rbp+3Ch],eax  
00007FFF013A183A  mov         r8d,dword ptr [rbp+3Ch]  
00007FFF013A183E  mov         dword ptr [rbp+54h],r8d  
            int length = ClampViaInEnd(in span, start, trimElement);
00007FFF013A1842  lea         r8,[rbp+90h]  
00007FFF013A1849  mov         rcx,qword ptr [rbp+88h]  
00007FFF013A1850  mov         edx,dword ptr [rbp+54h]  
00007FFF013A1853  call        00007FFF013A11E8  
00007FFF013A1858  mov         dword ptr [rbp+38h],eax  
00007FFF013A185B  mov         edx,dword ptr [rbp+38h]  
00007FFF013A185E  mov         dword ptr [rbp+50h],edx  
            return span.Slice(start, length);

            int start = ClampViaRefStart(ref span, trimElement); // Internal call stack uses "ref"
00007FFF013C1824  lea         rdx,[rbp+90h]  
00007FFF013C182B  mov         rcx,qword ptr [rbp+88h]  
00007FFF013C1832  call        00007FFF013C11C8  
00007FFF013C1837  mov         dword ptr [rbp+3Ch],eax  
00007FFF013C183A  mov         r8d,dword ptr [rbp+3Ch]  
00007FFF013C183E  mov         dword ptr [rbp+54h],r8d  
            int length = ClampViaRefEnd(ref span, start, trimElement);
00007FFF013C1842  lea         r8,[rbp+90h]  
00007FFF013C1849  mov         rcx,qword ptr [rbp+88h]  
00007FFF013C1850  mov         edx,dword ptr [rbp+54h]  
00007FFF013C1853  call        00007FFF013C11E8  
00007FFF013C1858  mov         dword ptr [rbp+38h],eax  
00007FFF013C185B  mov         edx,dword ptr [rbp+38h]  
00007FFF013C185E  mov         dword ptr [rbp+50h],edx  
            return span.Slice(start, length);
```

Reply:
For completeness, here's the assembly for the default passing style:
```
            int start = ClampViaNoneStart(span, trimElement);
00007FFF01381827  mov         rcx,qword ptr [rbp+0A8h]  
00007FFF0138182E  vmovdqu     xmm0,xmmword ptr [rcx]  
00007FFF01381833  vmovdqu     xmmword ptr [rbp+38h],xmm0  
00007FFF01381839  lea         rcx,[rbp+38h]  
00007FFF0138183D  mov         edx,dword ptr [rbp+0B0h]  
00007FFF01381843  movzx       edx,dl  
00007FFF01381846  call        00007FFF01381368  
00007FFF0138184B  mov         dword ptr [rbp+5Ch],eax  
00007FFF0138184E  mov         ecx,dword ptr [rbp+5Ch]  
00007FFF01381851  mov         dword ptr [rbp+74h],ecx  
            int length = ClampViaNoneEnd(span, start, trimElement);
00007FFF01381854  mov         rcx,qword ptr [rbp+0A8h]  
00007FFF0138185B  vmovdqu     xmm0,xmmword ptr [rcx]  
00007FFF01381860  vmovdqu     xmmword ptr [rbp+28h],xmm0  
00007FFF01381866  lea         rcx,[rbp+28h]  
00007FFF0138186A  mov         r8d,dword ptr [rbp+0B0h]  
00007FFF01381871  movzx       r8d,r8b  
00007FFF01381875  mov         edx,dword ptr [rbp+74h]  
00007FFF01381878  call        00007FFF01381388  
00007FFF0138187D  mov         dword ptr [rbp+58h],eax  
00007FFF01381880  mov         edx,dword ptr [rbp+58h]  
00007FFF01381883  mov         dword ptr [rbp+70h],edx  
            return span.Slice(start, length);
```
I am curious. Why does it use the SSE instructions here?